docs/architecture/2024-10-23_osquery_refactor.md

```diff
@@ -0,0 +1,92 @@
+# Osquery extension, runner, and instance refactor
+
+## Status
+
+2024-10-23: ADR submitted to team.
+
+## Context
+
+We want to be able to run between 0 and n osquery instances; currently, we run exactly one. In the short term, we want to be able to split up where Kolide queries run versus where tenant queries run, so that if a tenant creates a query that causes issues with osquery, the Kolide queries will still run. (This amounts to running 2 osquery instances.) In the longer term, we want to support multitenant installations. This will mean 1 osquery instance for Kolide queries, and 1 osquery instance per tenant.
+
+### Current implementation
+
+```mermaid
+---
+title: Current state
+---
+erDiagram
+    RUNGROUP ||--|| EXTENSION : runs
+    EXTENSION ||--|| JSONRPCCLIENT : uses
+    RUNGROUP ||--|| OSQUERYRUNNER : runs
+    OSQUERYRUNNER ||--|| OSQUERYINSTANCE: runs
+    OSQUERYINSTANCE ||--|{ EXTENSIONSERVER : runs
+    OSQUERYINSTANCE ||--|| EXTENSIONCLIENT : runs
+    EXTENSIONSERVER ||--|| EXTENSION : forwards
+    OSQUERYINSTANCE ||--|| STATS: owns
+    OSQUERYRUNNER ||--|| STATS : updates
+```
+
+In our current implementation, the extension is created by `runLauncher` and run via our rungroup, and passed into the osquery runner for use when creating the instances. To run multiple osquery instances using our _current_ implementation, we have two choices:
+
+1. We could share one osquery extension with all osquery instances
+2. We could create multiple extensions inside `runLauncher` and pass them into the runner to distribute one per osquery instance
+
+Ultimately, I rejected both of these options as unsuitable.
+
+We don't want to share the osquery extension among all osquery instances -- we will be unable to identify which osquery instance is talking to the extension, which makes this option untenable. Additionally, we may run into socket contention issues.
```

Comment:
Wow. This is a tough call. I find myself able to argue either side. 

In favor of sharing: Fewer go actors; less process/CPU overhead; less places for windows sleep to get in the way

In favor of split: More clear ownership and separation.

My hunch is socket contention may not be an issue -- I think with the go server side we can be more clever. Maybe? At least it's testable. 

I think it's _also_ worth thinking about the jsonrpc extension and the tables extensions a bit differently. Even if we decide it doesn't matter.

Reply:
The jsonrpc extension has some additional nuance. Do we want each osquery instance to talk to our SaaS via very distinct go processes, or do we want a single go processes that can redirect and distribute things. I highly suspect this will depends on how we handle auth flow.

I imagine we'll end up in a world where there's a set of root keys that identify launcher. And then we end up with some kind of registration data for each tenant. In that world, a shared extensions means a communication pattern where launcher root fetches everything, perhaps with each registration. In contrast, N extensions means each instance is using it's registration data to fetch.

It feels somewhat hard to predict. I think as long as we're prepared to rework it if we think we got it wrong, it doesn't matter which we try.

Reply:
> In favor of sharing: Fewer go actors; less process/CPU overhead; less places for windows sleep to get in the way

I've updated to add this to the `Consequences` section as well.